docarray/array/mixins/io/binary.py

```diff
@@ -41,37 +42,100 @@ def load_binary(
             file_ctx = open(file, 'rb')
         else:
             raise ValueError(f'unsupported input {file!r}')
+        if return_iterator:
+            return cls._load_binary_stream(
+                file_ctx, protocol=protocol, compress=compress
+            )
+        else:
+            return cls._load_binary_all(file_ctx, protocol, compress, _show_progress)
+
+    @classmethod
+    def _load_binary_stream(
+        cls: Type['T'], file_ctx: str, protocol=None, compress=None, show_progress=False
+    ) -> 'T':
 
         from .... import Document
 
+        if show_progress:
+            from rich.progress import track as _track
+
+            track = lambda x: _track(x, description='Deserializing')
+        else:
+            track = lambda x: x
+
+        with file_ctx as f:
+            version_numdocs_lendoc0 = f.read(9)
+            # 1 byte (uint8)
+            version = int.from_bytes(version_numdocs_lendoc0[0:1], 'big', signed=False)
+            # 8 bytes (uint64)
+            num_docs = int.from_bytes(version_numdocs_lendoc0[1:9], 'big', signed=False)
+
+            for _ in track(range(num_docs)):
+                # 4 bytes (uint32)
+                len_current_doc_in_bytes = int.from_bytes(
+                    f.read(4), 'big', signed=False
+                )
+                yield Document.from_bytes(
+                    f.read(len_current_doc_in_bytes),
+                    protocol=protocol,
+                    compress=compress,
+                )
+
+    @classmethod
+    def _load_binary_all(cls, file_ctx, protocol, compress, show_progress):
+        from .... import Document
+
         with file_ctx as fp:
             d = fp.read() if hasattr(fp, 'read') else fp
+
+        if protocol == 'pickle-array' or protocol == 'protobuf-array':
             if get_compress_ctx(algorithm=compress) is not None:
                 d = decompress_bytes(d, algorithm=compress)
                 compress = None
 
-            if protocol == 'protobuf-array':
-                from ....proto.docarray_pb2 import DocumentArrayProto
+        if protocol == 'protobuf-array':
+            from ....proto.docarray_pb2 import DocumentArrayProto
+
+            dap = DocumentArrayProto()
+            dap.ParseFromString(d)
 
-                dap = DocumentArrayProto()
-                dap.ParseFromString(d)
+            return cls.from_protobuf(dap)
+        elif protocol == 'pickle-array':
+            return pickle.loads(d)
 
-                return cls.from_protobuf(dap)
-            elif protocol == 'pickle-array':
-                return pickle.loads(d)
+        # Binary format for streaming case
+        else:
```

Comment:
I am not fan of this `else`. We should perhaps name it

Reply:
ok, making explicit the elif